Formula/leafnode.rb

```diff
@@ -14,8 +14,8 @@ def install
                           "--with-user=#{ENV["USER"]}", "--with-group=admin",
                           "--sysconfdir=#{etc}/leafnode", "--with-spooldir=#{var}/spool/news/leafnode"
     system "make", "install"
-    (prefix/"homebrew.mxcl.fetchnews.plist").write fetchnews_plist
-    (prefix/"homebrew.mxcl.texpire.plist").write texpire_plist
+    (prefix/"homebrew.mxcl.fetchnews.plist").write fetchnews_plist.to_plist
```

Comment:
I just did the same thing as with `def plist`… this formula happened to have custom property lists defined.

> Might be interesting to see if we can figure out a way to make this implicit?

@MikeMcQuaid Would that logic need to live in the `.write` method then? Or maybe the existing `def plist` method could handle mutliple property list {hash}es or something… what do you think?

Reply:
@danielbayley I was thinking maybe we should have a different method name e.g. `plist_hash` or similar and the `Formula` `plist*` methods could do the conversion to this format. What do you think?

Reply:
@MikeMcQuaid, To be honest, I'm thinking having more/seperate methods isn't ideal for maintainance… it feels like the ideal API would just be to have a single `def plist` method (or even just a `plist` stanza… why is it currently a method?) to handle either a {hash} or a string (maybe phasing the raw \<xml> string option out eventaully, as why would anyone want that given the alternative?), or an array of them (for formulas that need to create multiple plist files). The `.to_plist` logic is just a couple of lines of code after all. Also, how could it relate to `plist_name` and `plist_path`… maybe the `plist_name` could instead be taken from the `Label:` property (since they are supposed to match the file name anyway:
~~~ rb
plist [{
  Label: "homebrew.mxcl.one",
  Etc: "…",
},
{
  Label: "homebrew.mxcl.two",
  Etc: "…",
}]
~~~
What are your thoughts on these ideas?

Reply:
> To be honest, I'm thinking having more/seperate methods isn't ideal for maintainance

My main issue is that `.write` isn't going to fail but it just going to write out bad data. Changing a stable, public API such that it can return different data that needs to be handled differently Is increasingly seemingly like a bad idea to me.

The alternative could be that calling the plist "getter" methods always return `to_plist` such that you can _set_ a Hash but never _get_ a hash.

I'm game for eventually having an API that applies to all public formulae but you'll need to volunteer to be one of (or the only) person updating that.

> even just a `plist` stanza

This could be added as a new thing that only accepted a Hash.

---

Some formulae do want to have multiple plists so this doesn't seem like a bad idea. We'd need to figure out how `brew services` would handle that, though, as that's a/the major consumer of these plists (https://github.com/Homebrew/homebrew-services)

Reply:
> `.write` isn't going to fail but it just going to write out bad data.

Maybe `plutil -lint` could mitigate that, but yes messing with `.write` is probably a bad way to go about this.

> but you'll need to volunteer to be one of (or the only) person updating that.

My only real motivation here is that I want to put some formulas in a tap that write a [single] plist file each, and looking at how it's currently done, I can see that the DLS/API for this particular task could easily, and should be a lot nicer than it currently is. For my needs this is just being able to write a simple Ruby {hash} that generates the plist for `homebrew-services`… and all that really requires is the very small amount of (`.to_plist`) code already in https://github.com/Homebrew/brew/pull/5695. A nice benefit is that it can tidy up a lot of existing formulae, but I'm not going to realistically go thorough and update them all (this could probably be scripted actually… which might be the best approach?), and the formulae where there are multiple plists `def`ined I don't really use either… but there _should_ really be an easy way to just define either a single, or an array of hashes in general, to write out a couple of plist files.

> This could be added as a new thing that only accepted a Hash.

This feels like it could be the way to go, and maybe just put in the docs/Formula Cookbook that all future formula should use this `plist` stanza, along with an example, and just phase the old methods out… what do you think? @MikeMcQuaid

@colindean @reitermarkus care to chip in with opinions here?

Reply:
I'm thinking something like:

```ruby
def plist(label)
  @plists ||= {}

  if block_given?
    plist = yield
    plist[:Label] = label
    @plists[label] = plist.to_plist
  else 
    @plists[label]
  end
end
```

Then specify it like

```ruby
plist 'sh.brew.dnsmasq' do
  {
    ProgramArguments: %w[#{opt_sbin}/dnsmasq --keep-in-foreground -C #{etc}/dnsmasq.conf],
    RunAtLoad:        true,
    KeepAlive:        true,
  }
end
```

and get it with 
```ruby
plist 'sh.brew.dnsmasq'
```

Reply:
Works for me. As long as calling `Formula#plist` returns the same value regardless 👍

Reply:
@reitermarkus So how is this supposed to work exactly? I can't get the `def plist` method in `Library/Homebrew/formula.rb` to recognise the `do` block passed to it in a test `Formula/*.rb`… The `block_given?` condition never kicks in, and so doesn't `yield`. Also, I presume you wrote it this way to enable multiple `plist do` blocks to be defined in a single formula?

https://github.com/Homebrew/brew/pull/5695#issuecomment-465500282

> Feel free to ask for any questions/clarification on your pull requests and I'll do my best to help in a timely fashion.

@MikeMcQuaid Basically I just need to figure out how to impliment a stanza properly… (this will also be useful to know for my _[homebrew-alfred](https://github.com/danielbayley/homebrew-alfred)_ tap). Am I missing things defined in other places than `Library/Homebrew/formula.rb`? I suppose it doesn't help that the codebase seems very different between `brew` and `cask` either…

Reply:
> Basically I just need to figure out how to impliment a stanza properly…

@danielbayley What do you mean by "stanza"; something that like `plist` that defines a new class-wide Formula method?

> Am I missing things defined in other places than `Library/Homebrew/formula.rb`?

That should be the main place or at least main entry point. Where other classes are called they should be fairly obvious just from reading `formula.rb` but shout if they are not.

Reply:
> What do you mean by "stanza"; something that like `plist` that defines a new class-wide Formula method?

@MikeMcQuaid Maybe "stanza" is more of a `cask`-related term… but yes, basically. I get that currently, the `plist` method is meant to simply be overridden in a given formula with `def plist` (which is why in `formula.rb` it just contains `nil`)… but here we are trying to refactor it to accept a block (which passes in the {hash}), as well as it still being possible to re`def`ine, so as to not break existing formulae. The part that I'm having trouble with is what I outlined above:

> I can't get the `def plist` method in `Library/Homebrew/formula.rb` to recognise the `do` block passed to it in a test `Formula/*.rb`… The `block_given?` condition never kicks in, and so doesn't `yield`.

If I just replace
~~~ rb
# Library/Homebrew/formula.rb
def plist
  nil
end
~~~
with the method @reitermarkus suggested (or similar implimentation) then I get "undefined method `plist' for #<Class:0x0…>`" error, unless I also have a `def self.plist` method defined, or something like `attr_reader :plist`… but then it doesn't recieve the block, or `label` argument from the actual formula.

This seems like it should be simple on the surface, but this just reminds me why I stay away from OOP, `class` heirarchy and `self` (or `this` in JavaScript)…

Reply:
Yes, the `plist do` I suggested is `def self.plist`, while the one used by the formula is just `def plist`. So `plist` will have to forward the call to `self.plist`.

Reply:
Yeh, in general it's worth looking at the existing `plist` and other DSL code to see how it works and it requires both an instance and class method to be defined.

Reply:
@MikeMcQuaid @reitermarkus See the changes I pushed, and comment over at https://github.com/Homebrew/brew/pull/5695#issuecomment-468529524…